File: src/main/java/wooyoung/tom/simplespringboot/repository/JpaMemberRepository.java
package wooyoung.tom.simplespringboot.repository;

import wooyoung.tom.simplespringboot.domain.Member;

import javax.persistence.EntityManager;
import java.util.List;
import java.util.Optional;

public class JpaMemberRepository implements MemberRepository {

    // JPA 가 동작하려면 필요하다.
    private final EntityManager em;

    public JpaMemberRepository(EntityManager em) {
        this.em = em;
    }

    @Override
    public Member save(Member member) {
        // Insert Query 를 다 만들어 준다고?
        em.persist(member);
        return member;
    }

    @Override
    public Optional<Member> findById(Long id) {
        Member member = em.find(Member.class, id);
        return Optional.ofNullable(member);
    }

    @Override
    public Optional<Member> findByName(String name) {
        List<Member> result = em.createQuery("select m from Member m where m.name = :name", Member.class)
                .setParameter("name", name)
                .getResultList();

        return result.stream().findAny();
    }

    @Override
    public List<Member> findAll() {
        // 객체를 대상으로 SQL 수행하도록 한 것.
        return em.createQuery("select m from Member m", Member.class).getResultList();
    }
}

File: src/main/java/wooyoung/tom/simplespringboot/repository/member/JpaMemberRepository.kt
package wooyoung.tom.simplespringboot.repository.member

import wooyoung.tom.simplespringboot.domain.Member
import java.util.*
import javax.persistence.EntityManager

open class JpaMemberRepository(
    private val entityManager: EntityManager
) : MemberRepository {

    override fun save(member: Member): Member {
        entityManager.persist(member)
        return member
    }

    override fun findById(id: Long?): Optional<Member> {
        val member = entityManager.find(Member::class.java, id)
        return Optional.ofNullable(member)
    }

    override fun findByName(name: String?): Optional<Member> {
        return entityManager.createQuery("SELECT m from Member m WHERE m.name = :name", Member::class.java)
            .setParameter("name", name)
            .resultList.stream().findAny()
    }

    override fun findAll(): MutableList<Member> {
        return entityManager.createQuery("SELECT m FROM Member m", Member::class.java).resultList
    }
}